File: cpp/934.shortest-bridge.cpp
typedef pair<int, int> Point;

class Solution {
private:
    // find any land for a starting point
    Point findStart(vector<vector<int>>& grid) {
        auto m = grid.size();
        auto n = grid[0].size();
        for (auto i = 0; i < m; i++) {
            for (auto j = 0; j < n; j++) {
                if (grid[i][j] == 1) {
                    return make_pair(i, j);
                }
            }
        }
        return make_pair(0, 0);
    }

    // find all connected cells (whole island)
    // mark them as visited and put them into `vis`
    vector<Point> dfs(Point start, vector<vector<bool>>& vis, vector<vector<int>>& grid) {
        auto m = grid.size();
        auto n = grid[0].size();
        int phase[5] = {-1, 0, 1, 0, -1};
        vector<Point> res;

        stack<Point> st;
        st.push(start);
        vis[start.first][start.second] = true;

        while (!st.empty()) {
            auto node = st.top(); st.pop();
            res.push_back(node);

            for (int i = 0; i < 4; i++) {
                auto r = node.first + phase[i];
                auto c = node.second + phase[i + 1];

                if (0 <= r && r < m && 0 <= c && c < n && !vis[r][c] && grid[r][c]) {
                    vis[r][c] = true;
                    st.push(make_pair(r, c));
                }
            }
        }
        return res;
    }

    // find shortest path from visited island `island` to second island
    int bfs(vector<Point>& island, vector<vector<bool>>& vis, vector<vector<int>>& grid) {
        auto m = grid.size();
        auto n = grid[0].size();
        int phase[5] = {-1, 0, 1, 0, -1};
        
        // point coordinates (row, col) and it's distance from island
        queue<pair<Point, int>> que;
        // add all island coordinates with 0 distance
        for (const auto& p : island) {
            que.push(make_pair(p, 0));
        }

        while (!que.empty()) {
            auto [point, distance] = que.front(); que.pop();


            for (int i = 0; i < 4; i++) {
                auto r = point.first + phase[i];
                auto c = point.second + phase[i + 1];
                if (0 <= r && r < m && 0 <= c && c < n && !vis[r][c]) {
                    if (grid[r][c] == 1) {
                        return distance;
                    } else {
                        vis[r][c] = true;
                        que.push(make_pair(make_pair(r, c), distance + 1));
                    }
                }
            }
        }

        return -1;
    }

public:
    int shortestBridge(vector<vector<int>>& grid) {
        auto m = grid.size();
        auto n = grid[0].size();
        vector<vector<bool>> vis(m, vector<bool>(n, false));

        auto start = findStart(grid);
        auto island = dfs(start, vis, grid);

        return bfs(island, vis, grid);
    }
};

File: golang/934.shortest-bridge.go
type Point struct {
    y int
    x int
}

type Visited map[Point]bool

type Dist struct {
    p Point
    dist int
}

// find first land cell from one of two islands
func findStart(grid [][]int) Point {
    m, n := len(grid), len(grid[0])
    for i := 0; i < m; i++ {
        for j := 0; j < n; j++ {
            if grid[i][j] == 1 {
                return Point{i, j}
            }
        }
    }
    return Point{0, 0}
}

// detect all island cells and walk/mark them
func dfs(start Point, vis Visited, grid [][]int) []Point {
    m, n := len(grid), len(grid[0])
    res := make([]Point, 0, m * n)

    stack := make([]Point, 0, m * n)
    stack = append(stack, start)
    vis[start] = true
    phase := []int{-1, 0, 1, 0, -1}

    for len(stack) > 0 {
        node := stack[len(stack) - 1]
        stack = stack[:len(stack) - 1]
        res = append(res, node)

        for i := 0; i < 4; i++ {
            y, x := node.y + phase[i], node.x + phase[i + 1]
            nxt := Point{ y, x }
            if 0 <= y && y < m && 0 <= x && x < n && !vis[nxt] && grid[y][x] == 1 {
                vis[nxt] = true
                stack = append(stack, nxt)
            }
        }
    }

    return res
}

// start bfs simultaneously from 1st island
func bfs(start []Point, vis Visited, grid [][]int) int {
    m, n := len(grid), len(grid[0])
    queue := make([]Dist, 0, m * n)
    for _, p := range start {
        queue = append(queue, Dist{p, 0})
    }
    phase := []int{-1, 0, 1, 0, -1}

    for len(queue) > 0 {
        node := queue[0]
        queue = queue[1:]

        for i := 0; i < 4; i++ {
            y, x := node.p.y + phase[i], node.p.x + phase[i + 1]
            nxt := Point{ y, x }
            if 0 <= y && y < m && 0 <= x && x < n && !vis[nxt] {
                if grid[y][x] == 1 {
                    return node.dist
                } else {
                    vis[nxt] = true
                    queue = append(queue, Dist{nxt, node.dist + 1})
                }
            }
        }
    }

    return -1
}

func shortestBridge(grid [][]int) int {

    vis := make(Visited)

    start := findStart(grid)
    island := dfs(start, vis, grid)

    return bfs(island, vis, grid)
}